### PR TITLE
[rtl8721x] update rtl872x.tcl script for latest openocd version

### PR DESCRIPTION
**Pull request by hugomontero**
<details>
  <summary><i>submission notes</i></summary>

```
**Important:** Please sanitize/remove any confidential info like usernames, passwords, org names, product names/ids, access tokens, client ids/secrets, or anything else you don't wish to share.

Please Read and Sign the Contributor License Agreement ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md)).

You may also delete this submission notes header if you'd like. Thank you for contributing!
```
</details>

### Problem

What Bug or Missing Feature does this PR address? (A short summary is preferred over links)

### Solution

How does this PR address the problem stated above? (Describing the solutions implemented in code will facilitate a smooth discussion, review and testing of this PR)

### Steps to Test

Which unit/integration/application tests are applicable to this code change? (At minimum a test of some kind should be provided)

### Example App

```c
void setup() {
  /* A minimal example app is super helpful 
   * for testing new features and fixes. 
   * A link to a Docs PR is even better!
   */
}

void loop() {

}
```

### References

Links to the Community, Docs, Other Issues, etc..

---

### Completeness

- [x] User is totes amazing for contributing!
- [ ] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [ ] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

**Review by avtolstoy — APPROVED** [review]

File: scripts/rtl872x.tcl [hunk 1]
```diff
@@ -107,7 +108,8 @@ proc openocd_print_memory {address length} {
     init
     reset halt
     set value ""
-    mem2array value 8 $address $length
+		set value [read_memory $address 8 $length]
```
Comment (avtolstoy) [inline scripts/rtl872x.tcl:111]: Nit: spaces vs tabs. Same above.

File: scripts/rtl872x.tcl [hunk 2]
```diff
@@ -153,19 +155,20 @@ proc rtl872x_init {} {
         echo "initializing rtl872x"
         init
         reset halt
+				#write_memory 0x00082000 32 rtl872x_flash_loader 
         array2mem rtl872x_flash_loader 32 0x00082000 [array size rtl872x_flash_loader]
```
Comment (avtolstoy) [inline scripts/rtl872x.tcl:159]: This looks like the correct syntax. Is this not working?

File: scripts/rtl872x.tcl (hunk 1 above)
Comment (hugomontero) [reply scripts/rtl872x.tcl]: hehe fixed! My vim is not configured well hehe 😅

File: scripts/rtl872x.tcl (hunk 2 above)
Comment (hugomontero) [reply scripts/rtl872x.tcl:159]: I had problems with the progress bar when I did this change. 
After 100% of flashing it loops to 120..130... 150% and so on and so forth so idk why this behavior that's why I commented this instruction hehe.

**Comment by avtolstoy** [comment]
@hugomontero Looks like these commands are supported only since some commit on top of 0.11.0:
```
$ openocd --version
Open On-Chip Debugger 0.11.0 (2022-03-02-23:46)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
```

```
[Adapter 1] Unable to read device ID
initializing rtl872x
target halted due to debug-request, current mode: Thread
xPSR: 0x01000000 pc: 0x00000100 msp: 0x0008fffc
/home/avtolstoy/dev/device-os-2/scripts/target/rtl872x.tcl:97: Error: invalid command name "read_memory"
in procedure 'rtl872x_read_efuse_mac'
in procedure 'rtl872x_init' called at file "/home/avtolstoy/dev/device-os-2/scripts/target/rtl872x.tcl", line 391
in procedure 'openocd_read_register' called at file "/home/avtolstoy/dev/device-os-2/scripts/target/rtl872x.tcl", line 162
at file "/home/avtolstoy/dev/device-os-2/scripts/target/rtl872x.tcl", line 97>
```

File: scripts/rtl872x.tcl [hunk 3]
```diff
@@ -148,24 +150,26 @@ proc rtl872x_init {} {
...
     global RTL872x_REG_LP_KM4_CTRL
     global rtl872x_flash_loader
+    global rtl872x_flash_loader_data
 
     if {[expr {$rtl872x_ready == 0}]} {
         echo "initializing rtl872x"
         init
         reset halt
-        array2mem rtl872x_flash_loader 32 0x00082000 [array size rtl872x_flash_loader]
+        write_memory 0x00082000 32 rtl872x_flash_loader_data
```
Comment (hugomontero) [inline scripts/rtl872x.tcl:159]: @avtolstoy the problem you mentioned about "it stucks on init" it's because the old function `array2mem` have to convert the received array into an object and after that it calls to write_memory.
That's also why I couldn't just change the function call, because they receive different data types
See here 👉  https://review.openocd.org/c/openocd/+/6308/8/src/target/startup.tcl#222

I've create a new "flash_loader" with the expected object type and call directly the function and now it works ok.

File: scripts/rtl872x.tcl [hunk 4]
```diff
@@ -87,15 +87,16 @@ set rtl872x_ready                       0
...
+        return [expr {$address - $FLASH_MMU_START_ADDR}]
     }
     return $address
 }
 
 proc openocd_read_register {reg} {
     set value ""
-    mem2array value 32 $reg 1
-    return $value(0)
+    set value [read_memory $reg 32 1]
```
Comment (hugomontero) [inline scripts/rtl872x.tcl:97]: I'm pretty sure this one is going to fail. Because the way we're handling the value object in the follow lines.
Now the returned object is pretty different from the deprecated function

File: scripts/rtl872x.tcl [hunk 5]
```diff
@@ -107,7 +108,8 @@ proc openocd_print_memory {address length} {
     init
     reset halt
     set value ""
-    mem2array value 8 $address $length
+    set value [read_memory $address 8 $length]
```
Comment (hugomontero) [inline scripts/rtl872x.tcl:111]: I'm pretty sure this one is going to fail. Because the way we're handling the value object in the follow lines.
Now the returned object is pretty different from the deprecated function

**Comment by hugomontero** [comment]
> On Linux with toolchain installed openocd also doesn't work:
> 
> ```
> [Adapter 1] Target platform: rtl872x
> [Adapter 1] Starting OpenOCD; port: 4444
> [Adapter 1] $ openocd -f interface/cmsis-dap.cfg -f target/rtl872x.tcl -c "cmsis_dap_serial 10930000353891ab0533363639323946a5a5a5a597969906; telnet_port 4444; gdb_port disabled; tcl_port disabled; init"
> [Adapter 1] > rtl872x_read_efuse_mac; rtl872x_wdg_reset
> [Adapter 1] Stopping OpenOCD
> [Adapter 1] > shutdown
> [Adapter 1] response not received
> ```

I put a comment related with this problem here: https://github.com/particle-iot/device-os/pull/2525/files#r956261867
Now given the way you're testing, I'm sure it will stuck but now in flashing step. More exactly in verification step 👉  https://github.com/particle-iot/device-os/pull/2525/files#diff-5aab1178e8d5805ed5cebcf2daf19ed02f3f2c251c6b1e59406505ba76805b20R348

In order to test that, you can comment the verification part and you will see that everything will run as normal.

**Comment by scott-brust** [comment]
I pushed some changes to get things working with the new version of `jimtcl` which is `0.81` in particle toolchain 4 aka ` 0.11.0+dev (2022-03-25-19:34)`. 
This version uses the new `read_memory` / `write_memory` functions. I also hacked in some clunky way to make this script backwards compatible with `array2mem` / `mem2array` used in older versions of `jimtcl`. 
However, this version has a horrific regression in the speed in the `crc32` function. It takes about **400x** longer to calculate the CRC32 of a binary file on disk.

The default `brew install openocd` comes with a jimtcl version of `0.79`, which does not have an issue with the CRC32 speed.
The head build of `brew install --HEAD openocd` uses the same jimtcl version as the particle toolchain 4 (`0.81`), so the problem still exists in `jimtcl` and we cannot just update openocd versions...

tl;dr:
brew install openocd `0.11.0` = jimtcl 0.79 = no crc problem
particle toolchain4  `0.11.0+dev (2022-03-25-19:34)`= jimtcl 0.81 = very slow crc problem
latest openocd build `0.11.0+dev-00828-ga527112da (2022-08-30-21:01)` = jimtcl 0.81 = still very slow crc problem

**Review by keeramis — APPROVED** [review]
